### ISSUE TITLE
Consistent naming convention for Detekt CLI binaries

### ISSUE THREAD

**Issue by alyssa-glean** [detekt/detekt#4688]
## Expected Behavior
Ideally, the [install instructions](https://detekt.dev/cli.html#install-the-cli) for different platforms would result in a consistent state - a binary with a deterministic name that's on the user's `$PATH` and ready to use.

## Current Behavior
When installed via `brew`, the binary is access via `detekt [options]`. For Linux, the instructions given are to `curl` and `unzip` a bundle of JARs, resulting in a binary named `detekt-cli` (which, incidentally, is also not on the user's `$PATH`).

So I suppose I'm asking for two things:

1) In future releases, could the CLI binary be named simply `detekt` rather than `detekt-cli` to be consistent with the `brew` version?
2) Would it be appropriate to update the non-brew install instructions, perhaps to something like the below? I'd be happy to raise a PR if so, I've tested that this works on Ubuntu 18.04:

```
DETEKT_VERSION=1.19.0
curl -sSLO https://github.com/detekt/detekt/releases/download/v$DETEKT_VERSION/detekt-cli-$DETEKT_VERSION.zip &&
  unzip detekt-cli-$DETEKT_VERSION.zip &&
  chmod +x detekt-cli-$DETEKT_VERSION/bin/detekt-cli &&
  sudo ln -s $PWD/detekt-cli-$DETEKT_VERSION/bin/detekt-cli /usr/local/bin/detekt
```

Of course, you may have a preferred approach to going about this on Linux so happy to take guidance if that's the case. Perhaps we could get it into APT?

## Context
Our team use a mixture of MacOS and Linux, and I'm implementing a git hook that will invoke the detekt CLI on changed files (via `pre-commit`).

**Comment by cortinico** [comment]
> 1. In future releases, could the CLI binary be named simply `detekt` rather than `detekt-cli` to be consistent with the `brew` version?

I don't fully agree on this. I'm not sure if users are already depending already on this naming. Moreover now there is a 1-to-1 correspondence between the `:detekt-cli` Gradle module and the generated Jar.

I would argue that the one-off here is brew that is installing detekt essentially renaming the `detekt-cli.jar` to `detekt`.

> Would it be appropriate to update the non-brew install instructions, perhaps to something like the below? I'd be happy to raise a PR if so, I've tested that this works on Ubuntu 18.04:

Yup that's doable and we can totally update this.

**Comment by alyssa-glean** [comment]
> I would argue that the one-off here is brew that is installing detekt essentially renaming the `detekt-cli.jar` to `detekt`.

That's fair enough - I don't have any preference either way between which name becomes standard. Would it be possible to rename the brew version to `detekt-cli` going forwards?

**Comment by cortinico** [comment]
> Would it be possible to rename the brew version to `detekt-cli` going forwards?

What prevents you from renaming when installing on Linux?

**Comment by alyssa-glean** [comment]
> What prevents you from renaming when installing on Linux?

Nothing, of course - but we have a team of developers for whom we then have to maintain bespoke instructions somewhere. I felt it was worth making you aware of the inconsistency, as I'm sure we won't be the only cross-platform team to try and use the CLI. 

If there's a rename that's workable for you (one way or the other), you'd be helping to reduce friction for anyone using your tool. All other CLI tools that I can think of name themselves consistently across platforms - it just keeps things simpler.

**Comment by cortinico** [comment]
> I felt it was worth making you aware of the inconsistency, as I'm sure we won't be the only cross-platform team to try and use the CLI.

Thanks for pointing this out.

Sorry but I still fail to understand how renaming the brew formula will make things "easier" between cross-platform. 

For Linux, say you'r using a binary called `detekt-cli-1.19.0-all.jar` (so you'll `java -jar` that binary) 
For Mac, even if we rename the Brew formula to `detekt-cli`. You will still have two "inconsistent" naming. I don't see how `detekt` -> `detekt-cli` on Brew will simplify your setup.

Moreover, the Brew formula in not maintained by us. 

As for consistency, we've always kept the binary name consistent: `detekt-cli-<TOOL>-all.jar
https://github.com/detekt/detekt/releases/tag/v1.20.0-RC2.

Why don't you use the Jar for both Mac, Linux and Windows. I feel that Brew is adding unnecessary complexity to your setup.

**Comment by alyssa-glean** [comment]
My intention wasn't to invoke the jar on linux, but instead the provided binary `/bin/detekt-cli` (which you can invoke directly from a terminal without the java fluff :heart: ). 

And so, for the pre-commit hook, my WIP script is currently along the lines of: `detekt-cli --config <foo>...` - because I'm developing it on linux. But then I hit a snag - this won't work for Mac users who install via `brew`, because their binary is called `detekt` not `detekt-cli`. If these were consistent with each other, my hook would have "just worked" regardless of the install method/platform, reducing the chance of frustration in our team with broken hooks and/or confusing setup instructions. It would have been one less thing for me to think about.

> Why don't you use the Jar for both Mac, Linux and Windows. I feel that Brew is adding unnecessary complexity to your setup.

What you're suggesting here is that, for our use case, we should ignore your recommended install instructions (as well as what will be familiar/idiomatic for Mac users) and do our own thing. How are other adopters of your tool supposed to discover this going forwards, without first having to bump into a problem as I have? And if we do go down that route, we're again taking on a documentation burden. We can no longer just tell our developers to install the detekt CLI and link them to your install instructions, because that won't work for us - we have to tell them to do our own thing. It adds to the friction for folks who are onboarding.


So, to summarise: yes, there are ways I can work around this and move forward for my team - but I knew this when I opened the issue. My reason for raising was that this was an unexpected bump in the road I encountered when adopting your software, and one I thought might be straightforward to fix so others don't hit it going forwards.

**Comment by cortinico** [comment]
I understand your frustration, but I don't see what we can do to address this.

You can submit a new formula for `detekt-cli` and make sure the binary it installs is not renamed.

It's unlikely we will either:
- Rename the binary for everyone on Brew as that would be a major breaking change.
- Rename the binary inside the cli jar for the same reason

I think this was a mistake done by the author of the Brew formula (which is not us), but now it became the "standard". I think the easiest for you would be to either:

- Rename/Symlink the binary on Linux to just `detekt` as part of your setup.
- Use the `cli-all.jar` file on all the platforms (i.e. don't use Brew). We also mention "Any OS" on our Installation section https://detekt.dev/cli.html#install-the-cli. So in a sense, that's the most cross-platform compatible approach.

Fixing your issue now would introduce disruption for many other users, which I don't think it's something we're willing to do at this stage.

**Comment by alyssa-glean** [comment]
> Fixing your issue now would introduce disruption for many other users, which I don't think it's something we're willing to do at this stage.

Fair enough, that is the trade-off and it's your call. Thanks for taking the time to understand my issue :smile:
